File: internal/tui/theme.go
package tui

import (
	"image/color"

	"charm.land/lipgloss/v2"
)

type palette struct {
	panel    string
	promptBg string
	line     string
	line2    string
	ink      string
	muted    string
	faint    string
	faintest string
	accent   string
	green    string
	red      string
	amber    string
	blue     string
	onAccent string
	cardRun  string
	cardErr  string
}

type tuiTheme struct {
	ink             lipgloss.Style
	muted           lipgloss.Style
	faint           lipgloss.Style
	faintest        lipgloss.Style
	accent          lipgloss.Style
	green           lipgloss.Style
	red             lipgloss.Style
	amber           lipgloss.Style
	blue            lipgloss.Style
	line            lipgloss.Style
	lineStrong      lipgloss.Style
	badge           lipgloss.Style
	selectedRow     lipgloss.Style
	userPrompt      lipgloss.Style
	toolName        lipgloss.Style
	toolTarget      lipgloss.Style
	toolArg         lipgloss.Style
	cardRun         lipgloss.Style
	cardErr         lipgloss.Style
	bashPrompt      lipgloss.Style
	panel           lipgloss.Style
	userPromptPanel lipgloss.Style

	accentColor color.Color
	inkColor    color.Color
	bgPanel     color.Color
}

func buildTheme(p palette) tuiTheme {
	return tuiTheme{
		bgPanel:  lipgloss.Color(p.panel),
		ink:      lipgloss.NewStyle().Foreground(lipgloss.Color(p.ink)),
		muted:    lipgloss.NewStyle().Foreground(lipgloss.Color(p.muted)),
		faint:    lipgloss.NewStyle().Foreground(lipgloss.Color(p.faint)),
		faintest: lipgloss.NewStyle().Foreground(lipgloss.Color(p.faintest)),
		accent:   lipgloss.NewStyle().Foreground(lipgloss.Color(p.accent)),
		green:    lipgloss.NewStyle().Foreground(lipgloss.Color(p.green)),
		red:      lipgloss.NewStyle().Foreground(lipgloss.Color(p.red)),
		amber:    lipgloss.NewStyle().Foreground(lipgloss.Color(p.amber)),
		blue:     lipgloss.NewStyle().Foreground(lipgloss.Color(p.blue)),

		line:       lipgloss.NewStyle().Foreground(lipgloss.Color(p.line)),
		lineStrong: lipgloss.NewStyle().Foreground(lipgloss.Color(p.line2)),

		badge: lipgloss.NewStyle().
			Foreground(lipgloss.Color(p.onAccent)).
			Background(lipgloss.Color(p.accent)).
			Bold(true).
			Padding(0, 1),

		selectedRow: lipgloss.NewStyle().
			Foreground(lipgloss.Color(p.onAccent)).
			Background(lipgloss.Color(p.accent)).
			Bold(true),

		userPrompt: lipgloss.NewStyle().
			Foreground(lipgloss.Color(p.ink)).
			Bold(true),

		toolName: lipgloss.NewStyle().
			Foreground(lipgloss.Color(p.accent)).
			Bold(true),

		toolTarget: lipgloss.NewStyle().
			Foreground(lipgloss.Color(p.muted)),

		toolArg: lipgloss.NewStyle().
			Foreground(lipgloss.Color(p.faint)),

		cardRun: lipgloss.NewStyle().
			Foreground(lipgloss.Color(p.ink)).
			Background(lipgloss.Color(p.cardRun)),

		cardErr: lipgloss.NewStyle().
			Foreground(lipgloss.Color(p.ink)).
			Background(lipgloss.Color(p.cardErr)),

		bashPrompt: lipgloss.NewStyle().
			Foreground(lipgloss.Color(p.accent)).
			Bold(true),

		panel: lipgloss.NewStyle().
			Background(lipgloss.Color(p.panel)),

		userPromptPanel: lipgloss.NewStyle().
			Background(lipgloss.Color(p.promptBg)),

		accentColor: lipgloss.Color(p.accent),
		inkColor:    lipgloss.Color(p.ink),
	}
}

var zcTheme = buildTheme(darkPalette)

File: web/src/theme.ts
import { THEMES, type ThemeEntry, type Palette } from "./themes";

// Palette values are pushed onto :root as custom properties, so style.css
// names colours the same way internal/tui/theme.go does (ink, muted, faint,
// accent, line, ...) instead of hardcoding hex anywhere.
export function applyPalette(p: Palette) {
  const root = document.documentElement.style;
  root.setProperty("--panel", p.panel);
  root.setProperty("--prompt-bg", p.promptBg);
  root.setProperty("--line", p.line);
  root.setProperty("--line2", p.line2);
  root.setProperty("--ink", p.ink);
  root.setProperty("--muted", p.muted);
  root.setProperty("--faint", p.faint);
  root.setProperty("--faintest", p.faintest);
  root.setProperty("--accent", p.accent);
  root.setProperty("--green", p.green);
  root.setProperty("--red", p.red);
  root.setProperty("--amber", p.amber);
  root.setProperty("--blue", p.blue);
  root.setProperty("--on-accent", p.onAccent);
  root.setProperty("--card-run", p.cardRun);
  root.setProperty("--card-err", p.cardErr);
}

export function lookupTheme(name: string): ThemeEntry | undefined {
  const q = name.trim().toLowerCase();
  return THEMES.find((t) => t.name === q || t.label.toLowerCase() === q);
}

export function defaultTheme(): ThemeEntry {
  return THEMES[0]!;
}

export function applyTheme(name: string): ThemeEntry | undefined {
  const entry = lookupTheme(name);
  if (entry) {
    applyPalette(entry.palette);
    document.documentElement.style.setProperty("color-scheme", entry.isDark ? "dark" : "light");
  }
  return entry;
}
